oeps/oep-0006.rst

```diff
@@ -0,0 +1,224 @@
+
+For reference, the ``UserScope`` can have three values (no changes to these are
+being proposed):
+
+* ``NONE``: data not related any particular user, that does not change while
+  users interact with the XBlock
+* ``ALL``: data not related to any particular user, but which may be modified
+  during user interactions
+* ``ONE``: data related to an individual user
+
+The ``BlockScope`` can have the following values:
+
+* ``USAGE``: "The data is related to a particular use of a block in a course."
+* ``DEFINITION``: "The data is related to the definition of the block" (one block definition can be used multiple times in a course or even among multiple courses).
+* ``TYPE``: "The data is related to all instances of this type of XBlock."
+* ``ALL``: "The data is common to all blocks."
+
+XBlock runtimes would generally not be expected to implement every possible combination of ``BlockScope`` and ``UserScope``. Instead, certain "named scopes" are defined, which are the only scopes that most runtimes support:
+
+The existing named scopes, and an example use case for each, are as follows:
+
+* ``Scope.content`` (UserScope.NONE, BlockScope.DEFINITION):
+
+  * Example: The text of a multiple choice question
+* ``Scope.settings`` (UserScope.NONE, BlockScope.USAGE)
+
+  * Example: The number of attempts allowed for a problem
+* ``Scope.user_state`` (UserScope.ONE, BlockScope.USAGE)
+
+  * Example: The student's answer to a multiple choice question
+* ``Scope.user_state_summary`` (UserScope.ALL, BlockScope.USAGE)
+
+  * Example: The number of students who have answered the question (side note: there is no locking or atomicity for XBlock field writes, so in practice this field is currently of little use in a MOOC)
+* ``Scope.user_info`` (UserScope.ONE, BlockScope.ALL)
+
+  * Example: user age (note: this scope seems little-used and overlaps with the user service)
+* ``Scope.preferences`` (UserScope.ONE, BlockScope.TYPE)
+
+  * Example: (For a video XBlock) Boolean setting to prefer YouTube or non-YouTube video playback.
+
+
+Motivation
+==========
+
+As one would expect, ``Scope.user_info`` and ``Scope.preferences`` fields in the Open edX platform are implemented globally, and not scoped to the course that uses them. There is currently no standard, extensible way to use XBlock fields to store data that is scoped to a particular "context," such as a course, although this is a common need.
+
+
+Specification
+=============
+
+New BlockScope values
+---------------------
+
+To support context-scoped fields, ``BlockScope`` will allow these new additional values:
+
+* ``CONTEXT_TYPE``: The data is related to all instances of this type of XBlock within a particular context.
+* ``CONTEXT_ALL``: The data is accessible to XBlocks of any type, within a particular context. (As with ``BlockScope.ALL``, XBlock authors who use such fields should be careful to use a namespace that will not conflict with ``BlockScope.CONTEXT_ALL`` fields defined by other XBlock authors.)
```

Comment:
@bradenmacdonald: I think we need to do some more thinking about `CONTEXT_ALL` (a.k.a `CONTEXT_ALL_TYPES`).  

I was discussing with @andy-armstrong the [needs of UIBlocks and UIBlockTabs](https://openedx.atlassian.net/wiki/display/TNL/UIBlocks+for+Course+Outline) in the context of Discussions.  It seems clear that we need a way to share settings across UIBlocks and XBlocks (like the proposed `CONTEXT_ALL`), but it is unclear what the best mechanism would be that allows for clear ownership over the declaration and administration of these settings, as well as some mechanism for sharing these settings, even with UIBlocks.

While trying to define the differences between UIBlocks and XBlocks, we said UIBlocks would not be part of the content hierarchy.  The owner of these settings also doesn't need to be part of the content hierarchy, but shares functionality that XBlocks provide.  Is this a third type of block?  Is it just another XBlock with a specialized purpose?   Other?  This is just to kickstart additional thoughts on this topic.